### PR TITLE
ext_authz: Check for cluster before sending HTTP request

### LINKED ISSUE #8025: Envoy crashes when ext_authz is reconfigured while under load
*Title*: *Envoy crashes when ext_authz is reconfigured while under load*

*Description*:
>When a steady load of approximately > 10 requests per second is sent through Envoy, a reload of the `ext_authz` filter will cause envoy to appropriately return a 403 but then unexpectedly crash.

This has been tested on master as well as the v1.11.1 release.

*Repro steps*:
1. Stand up envoy with the `ext_authz` using a controller (in this case Ambassador)
2. Send a stream of requests through envoy
3. Reconfigure ext_authz
4. See envoy crash

*Config*:
```
{
    "@type": "/envoy.config.bootstrap.v2.Bootstrap",
    "static_resources": {
        "clusters": [
            {
                "connect_timeout": "3.000s",
                "dns_lookup_family": "V4_ONLY",
                "lb_policy": "ROUND_ROBIN",
                "load_assignment": {
                    "cluster_name": "cluster_127_0_0_1_8877",
                    "endpoints": [
                        {
                            "lb_endpoints": [
                                {
                                    "endpoint": {
                                        "address": {
                                            "socket_address": {
                                                "address": "127.0.0.1",
                                                "port_value": 8877,
                                                "protocol": "TCP"
                                            }
                                        }
                                    }
                                }
                            ]
                        }
                    ]
                },
                "name": "cluster_127_0_0_1_8877",
                "type": "STRICT_DNS"
            },
            {
                "connect_timeout": "3.000s",
                "dns_lookup_family": "V4_ONLY",
                "lb_policy": "ROUND_ROBIN",
                "load_assignment": {
                    "cluster_name": "cluster_extauth_example_auth2_3000",
                    "endpoints": [
                        {
                            "lb_endpoints": [
                                {
                                    "endpoint": {
                                        "address": {
                                            "socket_address": {
                                                "address": "example-auth2",
                                                "port_value": 3000,
                                                "protocol": "TCP"
                                            }
                                        }
                                    }
                                }
                            ]
                        }
                    ]
                },
                "name": "cluster_extauth_example_auth2_3000",
                "type": "STRICT_DNS"
            },
            {
                "connect_timeout": "3.000s",
                "dns_lookup_family": "V4_ONLY",
                "lb_policy": "ROUND_ROBIN",
                "load_assignment": {
                    "cluster_name": "cluster_tour_8080",
                    "endpoints": [
                        {
                            "lb_endpoints": [
                                {
                                    "endpoint": {
                                        "address": {
                                            "socket_address": {
                                                "address": "tour",
                                                "port_value": 8080,
                                                "protocol": "TCP"
                                            }
                                        }
                                    }
                                }
                            ]
                        }
                    ]
                },
                "name": "cluster_tour_8080",
                "type": "STRICT_DNS"
            }
        ],
        "listeners": [
            {
                "address": {
                    "socket_address": {
                        "address": "0.0.0.0",
                        "port_value": 8080,
                        "protocol": "TCP"
                    }
                },
                "filter_chains": [
                    {
                        "filters": [
                            {
                                "config": {
                                    "access_log": [
                                        {
                                            "config": {
                                                "format": "ACCESS [%START_TIME%] \"%REQ(:METHOD)% %REQ(X-ENVOY-ORIGINAL-PATH?:PATH)% %PROTOCOL%\" %RESPONSE_CODE% %RESPONSE_FLAGS% %BYTES_RECEIVED% %BYTES_SENT% %DURATION% %RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)% \"%REQ(X-FORWARDED-FOR)%\" \"%REQ(USER-AGENT)%\" \"%REQ(X-REQUEST-ID)%\" \"%REQ(:AUTHORITY)%\" \"%UPSTREAM_HOST%\"\n",
                                                "path": "/dev/fd/1"
                                            },
                                            "name": "envoy.file_access_log"
                                        }
                                    ],
                                    "http_filters": [
                                        {
                                            "config": {
                                                "http_service": {
                                                    "authorization_request": {
                                                        "allowed_headers": {
                                                            "patterns": [
                                                                {
                                                                    "exact": "x-qotm-session"
                                                                },
                                                                {
                                                                    "exact": "authorization"
                                                                },
                                                                {
                                                                    "exact": "cookie"
                                                                },
                                                                {
                                                                    "exact": "x-forwarded-for"
                                                                },
                                                                {
                                                                    "exact": "from"
                                                                },
                                                                {
                                                                    "exact": "x-forwarded-host"
                                                                },
                                                                {
                                                                    "exact": "x-forwarded-proto"
                                                                },
                                                                {
                                                                    "exact": "user-agent"
                                                                },
                                                                {
                                                                    "exact": "proxy-authorization"
                                                                }
                                                            ]
                                                        },
                                                        "headers_to_add": []
                                                    },
                                                    "authorization_response": {
                                                        "allowed_client_headers": {
                                                            "patterns": [
                                                                {
                                                                    "exact": "x-qotm-session"
                                                                },
                                                                {
                                                                    "exact": "set-cookie"
                                                                },
                                                                {
                                                                    "exact": "location"
                                                                },
                                                                {
                                                                    "exact": "www-authenticate"
                                                                },
                                                                {
                                                                    "exact": "authorization"
                                                                },
                                                                {
                                                                    "exact": "proxy-authenticate"
                                                                }
                                                            ]
                                                        },
                                                        "allowed_upstream_headers": {
                                                            "patterns": [
                                                                {
                                                                    "exact": "x-qotm-session"
                                                                },
                                                                {
                                                                    "exact": "set-cookie"
                                                                },
                                                                {
                                                                    "exact": "location"
                                                                },
                                                                {
                                                                    "exact": "www-authenticate"
                                                                },
                                                                {
                                                                    "exact": "authorization"
                                                                },
                                                                {
                                                                    "exact": "proxy-authenticate"
                                                                }
                                                            ]
                                                        }
                                                    },
                                                    "path_prefix": "/extauth",
                                                    "server_uri": {
                                                        "cluster": "cluster_extauth_example_auth2_3000",
                                                        "timeout": "5.000s",
                                                        "uri": "http://extauth"
                                                    }
                                                }
                                            },
                                            "name": "envoy.ext_authz"
                                        },
                                        {
                                            "name": "envoy.cors"
                                        },
                                        {
                                            "name": "envoy.router"
                                        }
                                    ],
                                    "http_protocol_options": {
                                        "accept_http_10": false
                                    },
                                    "normalize_path": true,
                                    "route_config": {
                                        "virtual_hosts": [
                                            {
                                                "domains": [
                                                    "*"
                                                ],
                                                "name": "backend",
                                                "routes": [
                                                    {
                                                        "match": {
                                                            "case_sensitive": true,
                                                            "prefix": "/ambassador/v0/check_ready",
                                                            "runtime_fraction": {
                                                                "default_value": {
                                                                    "denominator": "HUNDRED",
                                                                    "numerator": 100
                                                                },
                                                                "runtime_key": "routing.traffic_shift.cluster_127_0_0_1_8877"
                                                            }
                                                        },
                                                        "route": {
                                                            "cluster": "cluster_127_0_0_1_8877",
                                                            "prefix_rewrite": "/ambassador/v0/check_ready",
                                                            "priority": null,
                                                            "timeout": "10.000s"
                                                        }
                                                    },
                                                    {
                                                        "match": {
                                                            "case_sensitive": true,
                                                            "prefix": "/ambassador/v0/check_alive",
                                                            "runtime_fraction": {
                                                                "default_value": {
                                                                    "denominator": "HUNDRED",
                                                                    "numerator": 100
                                                                },
                                                                "runtime_key": "routing.traffic_shift.cluster_127_0_0_1_8877"
                                                            }
                                                        },
                                                        "route": {
                                                            "cluster": "cluster_127_0_0_1_8877",
                                                            "prefix_rewrite": "/ambassador/v0/check_alive",
                                                            "priority": null,
                                                            "timeout": "10.000s"
                                                        }
                                                    },
                                                    {
                                                        "match": {
                                                            "case_sensitive": true,
                                                            "prefix": "/ambassador/v0/",
                                                            "runtime_fraction": {
                                                                "default_value": {
                                                                    "denominator": "HUNDRED",
                                                                    "numerator": 100
                                                                },
                                                                "runtime_key": "routing.traffic_shift.cluster_127_0_0_1_8877"
                                                            }
                                                        },
                                                        "route": {
                                                            "cluster": "cluster_127_0_0_1_8877",
                                                            "prefix_rewrite": "/ambassador/v0/",
                                                            "priority": null,
                                                            "timeout": "10.000s"
                                                        }
                                                    },
                                                    {
                                                        "match": {
                                                            "case_sensitive": true,
                                                            "prefix": "/backend/",
                                                            "runtime_fraction": {
                                                                "default_value": {
                                                                    "denominator": "HUNDRED",
                                                                    "numerator": 100
                                                                },
                                                                "runtime_key": "routing.traffic_shift.cluster_tour_8080"
                                                            }
                                                        },
                                                        "route": {
                                                            "cluster": "cluster_tour_8080",
                                                            "prefix_rewrite": "/",
                                                            "priority": null,
                                                            "timeout": "3.000s"
                                                        }
                                                    }
                                                ]
                                            }
                                        ]
                                    },
                                    "server_name": "envoy",
                                    "stat_prefix": "ingress_http",
                                    "use_remote_address": true,
                                    "xff_num_trusted_hops": 0
                                },
                                "name": "envoy.http_connection_manager"
                            }
                        ],
                        "use_proxy_proto": false
                    }
                ],
                "name": "ambassador-listener-8080"
            }
        ]
    }
}
```

*Logs*:
```
[2019-08-23 14:04:48.715][198][trace][connection] [source/common/network/connection_impl.cc:394] [C1136] writing 178 bytes, end_stream false
[2019-08-23 14:04:48.715][195][debug][client] [source/common/http/codec_client.cc:82] [C885] disconnect. resetting 0 pending requests
[2019-08-23 14:04:48.715][195][debug][pool] [source/common/http/http1/conn_pool.cc:130] [C885] client disconnected, failure reason: 
[2019-08-23 14:04:48.715][195][trace][main] [source/common/event/dispatcher_impl.cc:158] item added to deferred deletion list (size=9)
[2019-08-23 14:04:48.715][195][debug][connection] [source/common/network/connection_impl.cc:101] [C272] closing data_to_write=0 type=1
[2019-08-23 14:04:48.715][195][debug][connection] [source/common/network/connection_impl.cc:190] [C272] closing socket: 1
[2019-08-23 14:04:48.715][195][debug][client] [source/common/http/codec_client.cc:82] [C272] disconnect. resetting 0 pending requests
[2019-08-23 14:04:48.715][195][debug][pool] [source/common/http/http1/conn_pool.cc:130] [C272] client disconnected, failure reason: 
[2019-08-23 14:04:48.715][195][trace][main] [source/common/event/dispatcher_impl.cc:158] item added to deferred deletion list (size=10)
[2019-08-23 14:04:48.715][195][debug][connection] [source/common/network/connection_impl.cc:101] [C274] closing data_to_write=0 type=1
[2019-08-23 14:04:48.715][195][debug][connection] [source/common/network/connection_impl.cc:190] [C274] closing socket: 1
[2019-08-23 14:04:48.715][195][debug][client] [source/common/http/codec_client.cc:82] [C274] disconnect. resetting 0 pending requests
[2019-08-23 14:04:48.715][195][debug][pool] [source/common/http/http1/conn_pool.cc:130] [C274] client disconnected, failure reason: 
[2019-08-23 14:04:48.715][195][trace][main] [source/common/event/dispatcher_impl.cc:158] item added to deferred deletion list (size=11)
[2019-08-23 14:04:48.715][195][debug][connection] [source/common/network/connection_impl.cc:101] [C863] closing data_to_write=0 type=1
[2019-08-23 14:04:48.715][195][debug][connection] [source/common/network/connection_impl.cc:190] [C863] closing socket: 1
[2019-08-23 14:04:48.715][195][debug][client] [source/common/http/codec_client.cc:82] [C863] disconnect. resetting 0 pending requests
[2019-08-23 14:04:48.715][195][debug][pool] [source/common/http/http1/conn_pool.cc:130] [C863] client disconnected, failure reason: 
[2019-08-23 14:04:48.715][195][trace][main] [source/common/event/dispatcher_impl.cc:158] item added to deferred deletion list (size=12)
[2019-08-23 14:04:48.715][195][trace][main] [source/common/event/dispatcher_impl.cc:158] item added to deferred deletion list (size=13)
[2019-08-23 14:04:48.715][195][trace][main] [source/common/event/dispatcher_impl.cc:76] clearing deferred deletion list (size=13)
[2019-08-23 14:04:48.715][195][trace][connection] [source/common/network/connection_impl.cc:458] [C1134] socket event: 3
[2019-08-23 14:04:48.715][195][trace][connection] [source/common/network/connection_impl.cc:543] [C1134] write ready
[2019-08-23 14:04:48.715][195][trace][connection] [source/common/network/connection_impl.cc:496] [C1134] read ready
[2019-08-23 14:04:48.715][195][trace][connection] [source/common/network/raw_buffer_socket.cc:23] [C1134] read returns: 284
[2019-08-23 14:04:48.715][195][trace][connection] [source/common/network/raw_buffer_socket.cc:37] [C1134] read error: Resource temporarily unavailable
[2019-08-23 14:04:48.715][195][trace][http] [source/common/http/http1/codec_impl.cc:375] [C1134] parsing 284 bytes
[2019-08-23 14:04:48.715][195][trace][http] [source/common/http/http1/codec_impl.cc:498] [C1134] message begin
[2019-08-23 14:04:48.715][195][trace][http] [source/common/http/http1/codec_impl.cc:343] [C1134] completed header: key=Content-Type value=application/json
[2019-08-23 14:04:48.715][198][trace][http] [source/common/http/conn_manager_impl.cc:1569] [C1136][S2563810555345383804] encode data called: filter=0x2dae0a0 status=0
[2019-08-23 14:04:48.715][191][debug][connection] [source/common/network/connection_impl.cc:101] [C1143] closing data_to_write=326 type=2
[2019-08-23 14:04:48.715][191][debug][connection] [source/common/network/connection_impl.cc:653] [C1143] setting delayed close timer with timeout 1000 ms
[2019-08-23 14:04:48.715][198][trace][http] [source/common/http/conn_manager_impl.cc:1582] [C1136][S2563810555345383804] encoding data via codec (size=144 end_stream=false)
[2019-08-23 14:04:48.715][195][trace][http] [source/common/http/http1/codec_impl.cc:343] [C1134] completed header: key=Date value=Fri, 23 Aug 2019 14:04:48 GMT
[2019-08-23 14:04:48.715][194][trace][connection] [source/common/network/raw_buffer_socket.cc:66] [C1144] write returns: 351
[2019-08-23 14:04:48.715][194][debug][connection] [source/common/network/connection_impl.cc:582] [C1144] write flush complete
[2019-08-23 14:04:48.715][194][trace][connection] [source/common/network/connection_impl.cc:458] [C1144] socket event: 6
[2019-08-23 14:04:48.715][194][debug][connection] [source/common/network/connection_impl.cc:479] [C1144] remote early close
[2019-08-23 14:04:48.716][194][debug][connection] [source/common/network/connection_impl.cc:190] [C1144] closing socket: 0
[2019-08-23 14:04:48.716][194][debug][main] [source/server/connection_handler_impl.cc:80] [C1144] adding to cleanup list
[2019-08-23 14:04:48.716][194][trace][main] [source/common/event/dispatcher_impl.cc:158] item added to deferred deletion list (size=1)
[2019-08-23 14:04:48.716][194][trace][connection] [source/common/network/connection_impl.cc:458] [C379] socket event: 3
[2019-08-23 14:04:48.716][194][trace][connection] [source/common/network/connection_impl.cc:543] [C379] write ready
[2019-08-23 14:04:48.716][194][trace][connection] [source/common/network/connection_impl.cc:496] [C379] read ready
[2019-08-23 14:04:48.716][194][trace][connection] [source/common/network/raw_buffer_socket.cc:23] [C379] read returns: 285
[2019-08-23 14:04:48.716][194][trace][connection] [source/common/network/raw_buffer_socket.cc:37] [C379] read error: Resource temporarily unavailable
[2019-08-23 14:04:48.716][194][trace][http] [source/common/http/http1/codec_impl.cc:375] [C379] parsing 285 bytes
[2019-08-23 14:04:48.716][194][trace][http] [source/common/http/http1/codec_impl.cc:498] [C379] message begin
[2019-08-23 14:04:48.716][194][trace][http] [source/common/http/http1/codec_impl.cc:343] [C379] completed header: key=Content-Type value=application/json
[2019-08-23 14:04:48.716][194][trace][http] [source/common/http/http1/codec_impl.cc:343] [C379] completed header: key=Date value=Fri, 23 Aug 2019 14:04:48 GMT
[2019-08-23 14:04:48.716][194][trace][http] [source/common/http/http1/codec_impl.cc:464] [C379] headers complete
[2019-08-23 14:04:48.716][194][trace][http] [source/common/http/http1/codec_impl.cc:343] [C379] completed header: key=Content-Length value=176
[2019-08-23 14:04:48.716][194][debug][router] [source/common/router/router.cc:1002] [C1145][S1885932356023518673] upstream headers complete: end_stream=false
[2019-08-23 14:04:48.716][194][trace][http] [source/common/http/conn_manager_impl.cc:1329] [C1145][S1885932356023518673] encode headers called: filter=0x2c66e10 status=0
[2019-08-23 14:04:48.716][194][debug][http] [source/common/http/conn_manager_impl.cc:1373] [C1145][S1885932356023518673] closing connection due to connection close header
[2019-08-23 14:04:48.716][194][debug][http] [source/common/http/conn_manager_impl.cc:1435] [C1145][S1885932356023518673] encoding headers via codec (end_stream=false):
':status', '200'
'content-type', 'application/json'
'date', 'Fri, 23 Aug 2019 14:04:48 GMT'
'content-length', '176'
'x-envoy-upstream-service-time', '1'
'server', 'envoy'
'connection', 'close'

[2019-08-23 14:04:48.716][194][trace][connection] [source/common/network/connection_impl.cc:394] [C1145] writing 177 bytes, end_stream false
[2019-08-23 14:04:48.716][194][trace][http] [source/common/http/conn_manager_impl.cc:1569] [C1145][S1885932356023518673] encode data called: filter=0x2c66e10 status=0
[2019-08-23 14:04:48.716][194][trace][http] [source/common/http/conn_manager_impl.cc:1582] [C1145][S1885932356023518673] encoding data via codec (size=176 end_stream=false)
[2019-08-23 14:04:48.716][194][trace][connection] [source/common/network/connection_impl.cc:394] [C1145] writing 176 bytes, end_stream false
[2019-08-23 14:04:48.716][194][trace][http] [source/common/http/http1/codec_impl.cc:485] [C379] message complete
[2019-08-23 14:04:48.716][194][trace][http] [source/common/http/http1/codec_impl.cc:760] [C379] message complete
[2019-08-23 14:04:48.716][194][debug][client] [source/common/http/codec_client.cc:95] [C379] response complete
[2019-08-23 14:04:48.716][194][trace][main] [source/common/event/dispatcher_impl.cc:158] item added to deferred deletion list (size=2)
[2019-08-23 14:04:48.716][194][trace][http] [source/common/http/conn_manager_impl.cc:1569] [C1145][S1885932356023518673] encode data called: filter=0x2c66e10 status=0
[2019-08-23 14:04:48.716][194][trace][http] [source/common/http/conn_manager_impl.cc:1582] [C1145][S1885932356023518673] encoding data via codec (size=0 end_stream=true)
[2019-08-23 14:04:48.716][194][trace][main] [source/common/event/dispatcher_impl.cc:158] item added to deferred deletion list (size=3)
[2019-08-23 14:04:48.716][194][debug][connection] [source/common/network/connection_impl.cc:101] [C1145] closing data_to_write=353 type=2
[2019-08-23 14:04:48.716][194][debug][connection] [source/common/network/connection_impl.cc:653] [C1145] setting delayed close timer with timeout 1000 ms
[2019-08-23 14:04:48.716][194][debug][pool] [source/common/http/http1/conn_pool.cc:199] [C379] response complete
[2019-08-23 14:04:48.716][194][debug][pool] [source/common/http/http1/conn_pool.cc:237] [C379] moving to ready
[2019-08-23 14:04:48.716][194][trace][http] [source/common/http/http1/codec_impl.cc:396] [C379] parsed 285 bytes
[2019-08-23 14:04:48.716][194][trace][main] [source/common/event/dispatcher_impl.cc:76] clearing deferred deletion list (size=3)
[2019-08-23 14:04:48.716][194][debug][main] [source/server/connection_handler_impl.cc:287] [C1164] new connection
[2019-08-23 14:04:48.716][194][trace][connection] [source/common/network/connection_impl.cc:458] [C1145] socket event: 2
[2019-08-23 14:04:48.716][194][trace][connection] [source/common/network/connection_impl.cc:543] [C1145] write ready
[2019-08-23 14:04:48.716][194][trace][connection] [source/common/network/raw_buffer_socket.cc:66] [C1145] write returns: 353
[2019-08-23 14:04:48.716][194][debug][connection] [source/common/network/connection_impl.cc:582] [C1145] write flush complete
[2019-08-23 14:04:48.716][194][trace][connection] [source/common/network/connection_impl.cc:458] [C1164] socket event: 3
[2019-08-23 14:04:48.716][194][trace][connection] [source/common/network/connection_impl.cc:543] [C1164] write ready
[2019-08-23 14:04:48.716][194][trace][connection] [source/common/network/connection_impl.cc:496] [C1164] read ready
[2019-08-23 14:04:48.716][194][trace][connection] [source/common/network/raw_buffer_socket.cc:23] [C1164] read returns: 238
[2019-08-23 14:04:48.716][194][trace][connection] [source/common/network/raw_buffer_socket.cc:37] [C1164] read error: Resource temporarily unavailable
[2019-08-23 14:04:48.716][194][trace][http] [source/common/http/http1/codec_impl.cc:375] [C1164] parsing 238 bytes
[2019-08-23 14:04:48.716][194][trace][http] [source/common/http/http1/codec_impl.cc:498] [C1164] message begin
[2019-08-23 14:04:48.716][194][debug][http] [source/common/http/conn_manager_impl.cc:247] [C1164] new stream
[2019-08-23 14:04:48.716][194][trace][http] [source/common/http/http1/codec_impl.cc:343] [C1164] completed header: key=Host value=192.168.39.204:30116
[2019-08-23 14:04:48.716][194][trace][http] [source/common/http/http1/codec_impl.cc:343] [C1164] completed header: key=Accept value=*/*
[2019-08-23 14:04:48.716][194][trace][http] [source/common/http/http1/codec_impl.cc:343] [C1164] completed header: key=Accept-Encoding value=gzip, deflate
[2019-08-23 14:04:48.716][194][trace][http] [source/common/http/http1/codec_impl.cc:343] [C1164] completed header: key=User-Agent value=Mozilla/5.0 (redhat-x86_64-linux-gnu) Siege/4.0.4
[2019-08-23 14:04:48.716][194][trace][http] [source/common/http/http1/codec_impl.cc:343] [C1164] completed header: key=Authorization value=Basic dXNlcm5hbWU6cGFzc3dvcmQ=
[2019-08-23 14:04:48.716][194][trace][http] [source/common/http/http1/codec_impl.cc:464] [C1164] headers complete
[2019-08-23 14:04:48.716][194][trace][http] [source/common/http/http1/codec_impl.cc:343] [C1164] completed header: key=Connection value=close
[2019-08-23 14:04:48.716][194][trace][http] [source/common/http/http1/codec_impl.cc:485] [C1164] message complete
[2019-08-23 14:04:48.716][194][debug][http] [source/common/http/conn_manager_impl.cc:620] [C1164][S697422797895800214] request headers complete (end_stream=true):
':authority', '192.168.39.204:30116'
':path', '/backend/get-quote/'
':method', 'GET'
'accept', '*/*'
'accept-encoding', 'gzip, deflate'
'user-agent', 'Mozilla/5.0 (redhat-x86_64-linux-gnu) Siege/4.0.4'
'authorization', 'Basic dXNlcm5hbWU6cGFzc3dvcmQ='
'connection', 'close'

[2019-08-23 14:04:48.716][194][debug][http] [source/common/http/conn_manager_impl.cc:1167] [C1164][S697422797895800214] request end stream
[2019-08-23 14:04:48.716][194][trace][filter] [source/extensions/filters/http/ext_authz/ext_authz.cc:83] [C1164][S697422797895800214] ext_authz filter calling authorization server
[2019-08-23 14:04:48.716][194][critical][main] [source/exe/terminate_handler.cc:13] std::terminate called! (possible uncaught exception, see trace)
[2019-08-23 14:04:48.716][194][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:69] Backtrace (use tools/stack_decode.py to get line numbers):
[2019-08-23 14:04:48.716][194][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #0: [0x12d45c8]
[2019-08-23 14:04:48.716][194][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #1: [0x12d44d9]
[2019-08-23 14:04:48.716][194][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #2: [0x14974c6]
[2019-08-23 14:04:48.716][194][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:81] Caught Aborted, suspect faulting address 0x22b8000000b5
[2019-08-23 14:04:48.716][194][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:69] Backtrace (use tools/stack_decode.py to get line numbers):
[2019-08-23 14:04:48.716][194][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #0: [0x7f60606d64b0]
[2019-08-23 14:04:48.717][194][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #1: [0x12d44d9]
[2019-08-23 14:04:48.717][194][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #2: [0x14974c6]
ActiveStream 0x2d62880, stream_id_: 697422797895800214, has_continue_headers_: 0, is_head_request_: 0, decoding_headers_only_: 0, encoding_headers_only_: 0
request_headers_: 
  ':authority', '192.168.39.204:30116'
  ':path', '/backend/get-quote/'
  ':method', 'GET'
  'accept', '*/*'
  'accept-encoding', 'pt-encoding4:04:48.717][181][debug][config] [source/extensions/filters/network/http_connection_manager/config.cc:351]       name: envoy.ext_authz
', 'gzip, deflate'
  'user-agent', 'Mozilla/5.0 (redhat-x86_64-linux-gnu) Siege/4.0.4'
  'authorization', 'Basic dXNlcm5hbWU6cGFzc3dvcmQ='
  'x-forwarded-for', '192.168.39.1'
  'x-forwarded-proto', 'http'
  'x-envoy-internal', 'true'
  'x-request-id', 'd984b0dc-92d6-4e5f-a00e-bcfe3f885f00'
request_trailers_: null
response_headers_: null
response_trailers_: null
&stream_info_: 
  StreamInfoImpl 0x2d629b8, protocol_: 1, response_code_: null, response_code_details_: null, health_check_request_: 0, route_name_: 
./entrypoint.sh: line 22: warning: run_pending_traps: bad value in trap_list[17]: 0x55f053dd8dd3
./entrypoint.sh: line 19: warning: run_pending_traps: bad value in trap_list[17]: 0x55f053dd8dd3
2019-08-23 14:04:48 AMBASSADOR INFO envoy exited with status 134
./entrypoint.sh: line 19: warning: run_pending_traps: bad value in trap_list[17]: 0x55f053dd8dd3
2019-08-23 14:04:48 AMBASSADOR INFO killing extant processes
kill: can't kill pid 181: No such process
./entrypoint.sh: line 192: warning: run_pending_traps: bad value in trap_list[17]: 0x55f053dd8dd3
[2019-08-23 14:04:48 +0000] [125] [INFO] Handling signal: term
time="2019-08-23T14:04:48Z" level=info msg=Done
./entrypoint.sh: line 19: warning: run_pending_traps: bad value in trap_list[17]: 0x55f053dd8dd3
2019-08-23 14:04:48 AMBASSADOR INFO shutting down (1)
```

### PR DESCRIPTION
**Pull request by dio**
Description:
This patch adds a cluster existence checking before sending the HTTP `check` request to the responsible cluster.

Risk Level: Low
Testing: Unit
Docs Changes: N/A
Release Notes: N/A
Fixes #8025 

cc. @containscafeine

Signed-off-by: Dhi Aurrahman <dio@tetrate.io>

### CODE REVIEW

File: source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc [hunk 1]
```diff
@@ -202,9 +202,20 @@ void RawHttpClientImpl::check(RequestCallbacks& callbacks,
...
-  request_ = cm_.httpAsyncClientForCluster(config_->cluster())
-                 .send(std::move(message), *this,
-                       Http::AsyncClient::RequestOptions().setTimeout(config_->timeout()));
+  const std::string& cluster = config_->cluster();
+
+  // It's possible that the cluster specified in the filter configuration no longer exists due to a
+  // CDS removal.
+  if (!cm_.get(cluster)) {
+    // TODO(dio): Add stats related to this.
+    ENVOY_LOG(debug, "ext_authz cluster '{}' does not exist", cluster);
```
Comment (gsagula) [inline source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc]: We will probably need to add tracing here too. I can update https://github.com/envoyproxy/envoy/pull/8142 after you merge it.

File: test/extensions/filters/common/ext_authz/ext_authz_http_impl_test.cc [hunk 2]
```diff
@@ -377,6 +379,17 @@ TEST_F(ExtAuthzHttpClientTest, CancelledAuthorizationRequest) {
   client_.cancel();
 }
 
+TEST_F(ExtAuthzHttpClientTest, NoCluster) {
```
[the remark this answers is not in the document]
Comment (dio) [reply test/extensions/filters/common/ext_authz/ext_authz_http_impl_test.cc]: Done.

File: source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc (hunk 1 above)
Comment (dio) [reply source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc]: 👍

**Comment by dio** [comment]
@gsagula thanks for the review. I have updated it. PTAL.

File: source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc [hunk 3]
```diff
@@ -202,9 +202,20 @@ void RawHttpClientImpl::check(RequestCallbacks& callbacks,
...
   }
 
-  request_ = cm_.httpAsyncClientForCluster(config_->cluster())
-                 .send(std::move(message), *this,
-                       Http::AsyncClient::RequestOptions().setTimeout(config_->timeout()));
+  const std::string& cluster = config_->cluster();
+
+  // It's possible that the cluster specified in the filter configuration no longer exists due to a
+  // CDS removal.
+  if (!cm_.get(cluster)) {
```
Comment (gsagula) [inline source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc]: If I remember correctly Envoy code favours explicit null pointer comparison. 
```
if (cm_.get(cluster) == nullptr) { ... }
```

Comment (dio) [reply source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc]: Done.

**Comment by gsagula** [comment]
@dio I think you will need to merge master.

**Comment by dio** [comment]
Anything else @gsagula?

**Comment by dio** [comment]
@mattklein123 when you have time could you help to take a look? Thank you!

**Pull request merged**